lib/puppet/property.rb

```diff
@@ -205,6 +206,22 @@ def change_to_s(current_value, newvalue)
     end
   end
 
+  # Format the given string and conditionally redact the provided interpolation variables, depending on if
+  # this property is sensitive.
+  #
+  # @note Because the default implementation of #is_to_s returns the current value as-is, it doesn't necessarily
+  #   return a string. For the sake of sanity we just cast everything to a string for interpolation so we don't
+  #   introduce issues with unexpected property values.
+  #
+  # @see String#format
+  # @param fmt [String] The format string to interpolate.
+  # @param args [Array<String>] One or more strings to conditionally redact and interpolate into the format string.
+  #
+  # @return [String]
+  def format(fmt, *args)
+    fmt % args.map { |arg| @sensitive ? "[redacted]" : arg.to_s }
+  end
```

Comment:
In general these APIs are pretty horrid. For example, as noted in the doc comment `#is_to_s` returns the current value unmodified, so it's not necessarily a string. To make this more fun `Puppet::Parameter.format_value_for_display` specifically handles non-string values, and `Property#change_to_s` calls that method with the output of `#is_to_s`, so we have functions indicating they return strings passing data to values that specifically handle values other than strings.

It's like Camelot - a silly place that perhaps we shouldn't visit.